cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java

```diff
@@ -371,13 +373,23 @@ public void error(Throwable cause) {
           Closeables.closeQuietly((Closeable) programRunner);
         }
       }
+
+      final ProgramRunCompletionDetails completionInfo = runtimeClientService.getProgramCompletionDetails();
```

Comment:
Can you please check what happens if message publishing is delayed while program is finishing (e.g. instance runtime service is down). There is this waiting and I am not sure if it happened before this line of code is executed.

Reply:
The runtimeClientService does a final poll of the local TMS and publish when [its `close()` method is called](https://github.com/cdapio/cdap/blob/fd63842d5e1b46b17914af203adbf807245d431e/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java#L355-L357). The `RuntimeClientService` is closed in the [surrounding `finally` block](https://github.com/cdapio/cdap/blob/fd63842d5e1b46b17914af203adbf807245d431e/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java#L310-L315).

With the current implementation, I was trying to get the status form the runtime client service before it is closed. I also found that when a pipeline is killed, there is a high chance that the status message is flushed during this final poll when the `RuntimeClientService` being closed. To ensure we always have the completion status, I've now moved the code block to get the status from the `RuntimeClientService` to the very end of the `DefaultRuntimeJob#run()` method, after all the core services are close.